index.bs

```diff
@@ -1616,27 +1673,37 @@ partial interface MLGraphBuilder {
     **Returns:** an {{MLOperand}}. The permuted or transposed N-D tensor. 
 </div>
 
-## MLGraph ## {#api-mlgraph}
-The {{MLGraph}} interface represents a compiled computational graph. A compiled graph once constructed is immutable and cannot be subsequently changed.
+## MLTensor ## {#api-mltensor}
+The {{MLTensor}} interface represents a resource of a multidimensional array of numbers with a shape.
 
 <script type=idl>
-dictionary MLInput {
-  required (MLBufferView or WebGLTexture or GPUTexture) data;
-  sequence<long> dimensions;
-};
+typedef (MLBufferView or WebGLTexture or GPUTexture) MLResource;
 
-dictionary MLOutput {
-  (MLBufferView or WebGLTexture or GPUTexture) data;
-  sequence<long> dimensions;
+[SecureContext, Exposed=Window]
+interface MLTensor {
+  // Construct a tensor from a resource
+  constructor(MLResource resource, optional sequence<long> dimensions);
+
+  // The tensor dimensions
+  sequence<long> dimensions();
+
+  // The underlying resource of the tensor
+  MLResource resource();
+
+  // Asynchronously download the result of a computation onto a typed array.
+  Promise<ArrayBufferView> data();
```

Comment:
According to the table of device types and the resource types, the `MLTensor` that downloads the result to `ArrayBufferView` probably needs to be separated from `MLTensor` for GPU resources (say `MLGPUTensor`). It is because that the GPU resources have their own data download methods. e.g.,

```js
interface MLTensor {
  constructor(ArrayBufferView buffer, optional sequence<long> dimensions);
  sequence<long> dimensions();
  Promise<ArrayBufferView> data();
};

typedef (MLResourceBufferView or WebGLTexture or GPUTexture) MLGPUResource;

interface MLGPUTensor {
  constructor(MLGPUResource resource, optional sequence<long> dimensions);
  sequence<long> dimensions();
  MLGPUResource resource();
};
```

Reply:
> probably needs to be separated from ones for GPU resources (say MLGPUTensor). It is because that the GPU resources have their own data download methods. 

If the tensor has a backing GPU resource, the execution output will be on that backing resource. The `await data()` call is only when the user wants to download the data from the GPU resource to a CPU buffer. 

If the user wants to make a copy from the tensor resource, they should do it themselves outside of the WebNN call.

Reply:
> If the tensor has a backing GPU resource, the execution output will be on that backing resource. The `await data()` call is only when the user wants to download the data from the GPU resource to a CPU buffer.

We probably need to coordinate with other Web API, e.g. if the resource is a `GPUBuffer`, WebGPU defines [`mayAsync`](https://gpuweb.github.io/gpuweb/#dom-gpubuffer-mapasync) and [`getMappedRange`](https://gpuweb.github.io/gpuweb/#dom-gpubuffer-getmappedrange) for the data downloading.

My another question is if the tensor is created with an `ArrayBufferView`, what the `MLResource resource()` would return? If it returns the `ArrayBufferView`, what's the semantic difference from `Promise<ArrayBufferView> data()`?

Reply:
> My another question is if the tensor is created with an ArrayBufferView, what the MLResource resource() would return? If it returns the ArrayBufferView, what's the semantic difference from Promise<ArrayBufferView> data()?

`resource()` returns the backing resource of the `MLTensor` while `await data()` asynchronously download and layout-convert the content to a standard layout to a separate buffer. I believe you proposed this semantic for the WASM backend when an operation is executed stand alone.

Reply:
> We probably need to coordinate with other Web API, e.g. if the resource is a GPUBuffer, WebGPU defines mayAsync and getMappedRange for the data downloading.

We define `await data()` as a contract to asynchronously download the content to a CPU buffer. How exactly that download is implemented from a GPU resource to a CPU buffer is up to the implementor of this contract. 

If the caller of WebNN wants to make a copy from one GPU resource to the other, then that is entirely outside the scope of WebNN.

Reply:
> `resource()` returns the backing resource of the `MLTensor` while `await data()` asynchronously downloads and layout-converts the content to a standard layout onto a separate buffer. I believe you proposed this semantic for the WASM backend when an operation is executed stand alone.

If the `MLTensor` is created with an `ArrayBufferView`, e.g. a view of the existing Wasm memory, I suppose `await data()` should download and relayout the results to the given buffer instead of creating a separate one.

For example, in the following code snippet, the `preallocatedBuffer`, `backingBuffer` and `returnedBuffer` should be the same buffer, and `await outputTensor.data()` downloads and relayouts the results to that buffer. Is my understanding correct?
```js
const preallocatedBuffer = new Float32Array(sizeOfShape([3, 3]);
const outputTensor = new MLTensor(preallocatedBuffer);
graph.compute(inputs, {'output': outputTensor});
const backingBuffer = outputTensor.resource();
// backingBuffer === preallocatedBuffer and the results are not available in preallocatedBuffer.
const returnedBuffer = await outputTensor.data();
// returnedBuffer === preallocatedBuffer and the results are available in preallocatedBuffer.
```

Reply:
So according to this semantic, the assumptions would be:
1. `backingBuffer` and `preallocatedBuffer` are same object
2. `returnedBuffer` is different from `preallocatedBuffer`
3. before `await data()`, the results are not available in `preallocatedBuffer`
4. after `await data()`, the results are available in `returnedBuffer` and `preallocatedBuffer`

Is my understanding correct? 

Then, the `returnedBuffer` would not be used by the Wasm backend, because it needs another copy to move the results of `returnedBuffer` into its existing memory. The Wasm backend provides the `preallocatedBuffer` of its existing memory and would expect the results are downloaded and layout converted into that buffer.

Actually, to meet the requirement that downloads or layout converts into a pre-allocated buffer, how about overriding `MLTensor.data()` to accept an `ArrayBufferView` as a target, say

```js
interface MLTensor {
  constructor(ArrayBufferView buffer, optional sequence<long> dimensions);
  sequence<long> dimensions();
  Promise<ArrayBufferView> data();
  Promise<void> data(ArrayBufferView buffer)
};
```

Then,
```js
const preallocatedBuffer = new Float32Array(sizeOfShape([3, 3]);
const outputs = graph.compute(inputs);
await outputs.c.data(preallocatedBuffer);
// the results are downloaded and layout converted into preallocatedBuffer.
```

Reply:
My another question is about `MLTesnor.resource()`, what's the usage for the context created with `MLContextOptions`?

According to the "device types and the resource types" table, for the device type "default", "graphics", "compute" or "software", the only supported resource type is `ArrayBufferView`. It means the `MLTensor.resource()` would return an `ArrayBufferView`. However it seems not useful because the user code would eventually call `await data()`.

Did I miss anything?

Reply:
@huningxin You're giving 2 requirements that are a bit contradicting. They need to be reconciled before this design can go on. 

On one hand, you said that the operation-specific calls from TF.js would need to stage its intermediate result in the output tensor after the compute call, and that the subsequent download and layout conversion should happen as an explicit step afterward.

On another hand, you also said here that the WASM backend wants to avoid additional copy, would try to avoid a separate layout conversion step, and would want layout conversion to happy in-place. 

I personally do not prefer this 2-stage design as it complicates the API semantic for the compute step quite a bit. It's confusing to have many overload methods that are actually carrying out different API semantics and not simply extending the existing semantic. 

I need to think about this some more as it's really messy now. I would like the compute semantic that can be defined as clearly as possible so that the caller won't end up depending on the implementation-specific behaviors full of caveats and assumptions.

Reply:
This does look like a complex design issue. On one hand, we have data located in different devices, and the need to copy them. On the other hand, we also have asynchrony, which requires some mechanism to indicate when data is available (handled by "Promise" here). Should we separate these two aspects (instead of bundling them together for a specific usage)? Here are a few related points: (a) How do users reuse the same underlying memory (in a GPU, for example) over multiple calls? Do they create distinct MLTensors for each use? (b) How would users know if the previous use of some memory/buffer is now complete and that the memory is free to be reused? (c) Do we want the ability to pass the asynchronous data ("Promise") produced by one execution to be passed in as input to another execution, without creating extra copy/wait?

Reply:
This design discussion merits its own issue.

@wchao1115 mind opening one?

Reply:
Since this PR hasn't been merged, I don't think a new issue is appropriate. I think we should resolve it either in this PR (if it pulls through) or as a new PR (if this is abandoned).

Reply:
The convention is to document self-contained design discussions in their own issues. It helps discoverability and future generations who look for information on a certain design choice.

If you feel it is better to have these discussions (soon 50 comments) here in this PR e.g. due to complex interactions and to retain context that’s fine.

This PR can fix multiple issues at the same time, if you opt in to the idea.

Reply:
@wchao1115 , let me try to clarify

> On one hand, you said that the operation-specific calls from TF.js would need to stage its intermediate result in the output tensor after the compute call, and that the subsequent download and layout conversion should happen as an explicit step afterward.

Yes. This is good for performance. The implementation can keep the data on device and optimal internal layout as long as possible. `MLTensor` could be passed between graph computes. The implementation only needs to do data download and layout convert upon `await data()`

> On another hand, you also said here that the WASM backend wants to avoid additional copy, would try to avoid a separate layout conversion step, and would want layout conversion to happen in-place in the output buffer.

It's only about the target buffer of the results. The Wasm backend wants to download the results in standard layout to a pre-allocated buffer (usually the Wasm memory) instead of getting a newly returned buffer. Otherwise, Wasm backend needs to copy it back to its memory. 

For example, with current API, Wasm backend needs to copy the data of `returnedBuffer` to `wasmMemoryView` by `wasmMemoryView.set`.
```js
const wasmMemoryView = new Float32Array(wasmMemory, offset, length);
const outputs = graph.compute(inputs);
const returnedBuffer = await outputs.c.data();
wasmMemoryView.set(returnedBuffer );  // memory copy
```

If `data()` accepts the pre-allocated buffer, as I proposed, it would avoid that copy, for example
```js
const wasmMemoryView = new Float32Array(wasmMemory, offset, length);
const outputs = graph.compute(inputs);
await outputs.c.data(wasmMemoryView );
```

I don't think the requirement that downloads the results to a pre-allocated buffer conflicts with the separate data download and layout convert step. They are both useful for effective Wasm backend integration.

Reply:
@wchao1115 

> I need to think about this some more as it's really messy now. I would like the compute semantic that can be defined as clearly as possible so that the caller won't end up depending on the implementation-specific behaviors full of caveats and assumptions.

Agree. 

If you look at the device types and resource types table in this PR (I am copying it here for reference):

<table>
  <tr><th>Device Type<th>ArrayBufferView<th>GPUBuffer<th>GPUTexture<th>WebGLBuffer<th>WebGLTexture
  <tr><td>GPUDevice<td>Yes<td>Yes<td>Yes<td>No<td>No
  <tr><td>WebGLRenderingContext<td>Yes<td>No<td>No<td>Yes<td>Yes
  <tr><td>default<td>Yes<td>No<td>No<td>No<td>No
  <tr><td>graphics<td>Yes<td>No<td>No<td>No<td>No
  <tr><td>compute<td>Yes<td>No<td>No<td>No<td>No
  <tr><td>software<td>Yes<td>No<td>No<td>No<td>No
</table>

the `ArrayBufferView` is supported by all device types. On the other hand, the GPU resource types are context specific and may need coordination with other corresponding GPU APIs. That's the reason I propose to separate `MLTensor` for `ArrayBufferView` from GPU resource types. The `MLTensor` for `ArrayBufferView` would make this spec self-contained.

My refined proposal of `MLTensor` and `compute` API is:
```js
interface MLTensor {
  constructor(ArrayBufferView buffer, optional sequence<long> dimensions);
  sequence<long> dimensions();
  Promise<ArrayBufferView> data(); // return a newly allocated buffer containing the results.
  Promise<void> data(ArrayBufferView buffer); // download the results to pre-allocated buffer.
};

typedef record<DOMString, MLTensor> MLNamedTensors;

interface MLGraph {
  MLNamedTensors compute(MLNamedTensors inputs);
};
```

For the GPU resources, I think the requirement is a binding mechanism for compute inputs/outputs. So I suppose the [`MLInput`](https://webmachinelearning.github.io/webnn/#dictdef-mlinput) and [`MLOutput`](https://webmachinelearning.github.io/webnn/#dictdef-mloutput) of the current spec are still good for this purpose. Probably we can make it more GPU resources specific, e.g.,

```js
typedef  (WebGLBuffer or GPUBuffer or WebGLTexture or GPUTexture) MLGPUResource;

dictionary MLGPUInput {
  required MLGPUResource resource;
  sequence<long> dimensions;
};

dictionary MLGPUOutput {
  MLGPUResource resource;
  sequence<long> dimensions;
};

typedef record<DOMString, MLGPUInput> MLNamedGPUInputs;
typedef record<DOMString, MLGPUOutput> MLNamedGPUOutputs;

partial interface MLGraph {
  (MLNamedGPUOutputs or undefined) compute(MLNamedGPUInputs inputs, optional MLNamedGPUOutputs outputs = {});
};
```
The `compute` can be sync as well, because now the data downloading to `ArrayBufferView` are handled by `MLTensor`. For these GPU resources, there are corresponding GPU APIs to download the data asynchronously.

Any thoughts?

Reply:
Let's step back a bit and really think about this requirement for a sec:

> The framework may compute multiple ops without accessing the intermediate results. So if we allow passing the output of a compute to the next compute without making the data available, it would help the underlying implementation hide the additional copies and conversions. The idea is illustrated by following pseudo framework integration code:

```javascript
// assume conv2d_op, add_op, relu_op are compiled single-op graphs
// input = tf.tensor(values);
// conv2d = tf.conv2d(input, filter);
conv2d.mlOutput = conv2d_op.compute({'input': {resource: input.arrayBuffer}).output; // webnn backend
// add = tf.add(conv2d, bias);
add.mlOutput = add_op.compute({'a': conv2d.mlOutput}).output; // webnn backend
// relu = tf.relu(add);
relu.mlOutput = relu_op.compute({'input': add.mlOutput}).output;  // webnn backend
// await relu.data();
await relu.mlOutput.readDataAsync({resource: relu.arrayBuffer});  // webnn backend
```

i.e. if a framework graph `[input]->conv1->conv2->[output]` has each of the two `conv` operations compiled separately on the backend, we want the result out of `conv1` to be an opaque handle with implementation-specific layout format that can later be "fixed" in a final async call (e.g. `await relu.data()` in the pseudo code above).

On a second thought, this requirement can't be allowed. No matter where a convolution operation is defined, whether at the framework level or at webnn backend, an output of a public operation must be in a known layout format. 

While I understand the motive of potentially reducing copies at the backend, from the API standpoint, it simply cannot be allowed or we'll end up with a spec that says that the layout of an output of an operation is in an unknown hardware-specific format. 

This means that for a framework graph `[input]->conv1->conv2->[output]`, if a user (of the framework) executes the entire graph in one atomic action (load-and-run), the framework backend must compile the two convolution ops together in order to achieve the optimization that would allow the intermediate result between the two ops to stay in a native format. And that only the final output after `conv2` is converted back to a known layout format. This is precisely the value of graph compilation -- to optimize by hiding the intermediate results within the graph that aren't exposed to the user.

For the case where the framework graph is actually exposed to the user and that the user chooses to execute each op individually (eager execution), the execution result from each op must be produced and converted into a known layout format every step of the way. Yes, it won't be as efficient but that's exactly the trade-off the user is making when using eager execution -- to favor inspection of  the output of each operator in the graph over end-to-end graph execution performance.

Reply:
@wchao1115 

> whether at the framework level or at webnn backend, an output of a public operation must be in a known layout format.

Agree. So the result of `MLTensor.data()` is always the standard layout format.

> we'll end up with a spec that says that the layout of an output of an operation is in an unknown hardware-specific format.

The internal format is up to the implementation, the spec won't define that.

> For the case where the framework graph is actually exposed to the user and that the user chooses to execute each op in the graph individually (eager execution), the execution result from each op must be produced and converted into a known layout format every step of the way.

The conversion would only be needed when user code calls `MLTensor.data()`. Otherwise, it can be kept in the internal format and be consumed by the next op compute as an input. 

As a summary, if user codes wants to access the output data, call `MLTensor.data()`. The result is always in the standard layout format.

Reply:
> Agree. So the result of `MLTensor.data()` is always the standard layout format.

Not only that. The output of `MLGraphBuilder.conv2d` must also be in standard format. See detail in the [conv2d API spec](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-conv2d), both the size and the shape of the output operand are clearly defined, because this operation might be the last in the graph, and so its output in that case is the graph's output, which must be defined.

Reply:
> The output of `MLGraphBuilder.conv2d` must also be in standard format. See detail in the [conv2d API spec](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-conv2d), both the size and the shape of the output operand are clearly defined, because this operation might be the last in the graph, and so its output in that case is the graph's output, which must be defined.

Now with `MLTensor` as the graph output, the user code can only access the results by calling `MLTensor.data()`. So if `MLTensor.data()` always returns the standard layout format, it would not break the spec definition.

Did I miss anything?

Reply:
The output of the `conv2d` will be in the output buffer that is bound to the graph output in the `compute` call. Before this PR, there was the `MLOutput` dictionary where the `data` field holds the output buffer. In this PR, that would be the `resource()` call. This is true whether the output buffer is a GPU resource or an ArrayBufferView. I now think it's probably best to leave the async download step up to the caller, and not tying it to the WebNN API since the caller would know how to deal with data downloads from a resource type they specify. As for the operation itself, the output buffer needs to always be in standard formats to keep the spec well defined.